src/GraphQL/Introspection/__InputValue.cs

```diff
@@ -36,7 +36,7 @@ public __InputValue(bool allowAppliedDirectives = false)
                     if (hasDefault?.DefaultValue == null)
                         return null;
 
-                    var ast = hasDefault.DefaultValue.AstFromValue(context.Schema, hasDefault.ResolvedType);
+                    var ast = hasDefault.ResolvedType.ToAST(hasDefault.DefaultValue);
```

Comment:
How long have I disliked this piece of code. 👍

Reply:
In my previous comments, I pointed out that `ToAST` method can yield NRE. I understand that on practice this call preceded by `if (value != null)`, but in general case `ToAST` should not yield NRE.

Reply:
I don't mind if we change all the ToAST to prevent NRE, but it is not necessary.  `Serialize`, `ParseLiteral`, `ParseValue` and `ToAST` do not accept `null` as a valid argument and will always throw an error (or return null, which is explicitly defined as an error).  NRE is more explicit as it will indicate where the problem lies, versus returning `null` which is an undefined error.  Better yet would be `ArgumentNullException`.

Perhaps we should add a note to the xml comments that the argument for those methods cannot be null.

Another point to consider: ToAST could return `NullValue` if passed null.  But since `Serialize` does not accept `null`, it does not seem that `ToAST` should either.

Sometime it might be nice to change all of this so that null values ARE handled by `Serialize`, `ParseLiteral` and `ParseValue` -- so a null external value can have a different internal representation, such as an empty string.

Reply:
> (or return null, which is explicitly defined as an error)

I see many scalars that do so

> Perhaps we should add a note to the xml comments that the argument for those methods cannot be null.

Good case for NRT.

> Another point to consider: ToAST could return NullValue if passed null. But since Serialize does not accept null, it does not seem that ToAST should either.

I though about it. Maybe. I'm not sure.

> Sometime it might be nice to change all of this so that null values ARE handled by Serialize, ParseLiteral and ParseValue -- so a null external value can have a different internal representation, such as an empty string.

Agree. My point is that all public `ScalarGraphType` methods should ideally be written so as not to throw exceptions, and certainly not NRE.

Reply:
Sometimes an explicit error is better than returning null. It allows for examining the underlying (inner) error type to determine the cause of the problem. Returning null just causes the caller to throw a nondescript error. I might argue that Serialize should ALWAYS throw an exception for an invalid value unless it is an invalid type, in which case it should return null. I might also argue that ArgumentNullException should ALWAYS be thrown if value is null because it is a distinct programming error within the execution strategy. It is not a failed conversion. It’s a programming error outside the method. (As the method is currently defined)

Reply:
I’m thinking that we should issue another PR immediately after this one to have scalars handle null values. People have asked for it before, and with all the cleanup we have done for v4, it will be very easy to implement (finally!).

Reply:
OK.

Reply:
#2333